### ISSUE TITLE
User-defined types (UDTs)

### ISSUE THREAD

**Issue by hampuskraft** [Intreecom/scyllapy#34]
First of all, thanks for your efforts in creating this driver!

This library is everything I've been looking for, but I would love to see support for UDTs, which I use extensively. The official Python driver for Scylla/Cassandra makes working with UDTs notoriously tricky—they require you to either register the UDTs with compatible classes beforehand or manually convert values back and forth from `SimpleNamespace`.

Ideally, I'd like scyllapy to treat UDTs as PyDicts when converting from CQL to Python. I did make some attempts at getting this working, but I ran into a marshaling error (not enough bytes) when attempting to bind an insert query with a UDT in the tests. If I understand the code correctly, the serialization to a Scylla-compatible type will happen in `serialize` in `ScyllaPyCQLDTO`. I attempted to implement a `UserDefinedType` that worked similarly to `Map`, but no dice.

https://github.com/Intreecom/scyllapy/blob/863da3967c1478ea5fe7a3961d3eb236a34b69af/src/utils.rs#L465-L468

I would love to hear your thoughts on the feasibility of implementing UDTs treated as dictionaries (both in results and params) without requiring registering them beforehand. I'd love to help, but you know the code better than I do. Thanks!

**Comment by s3rius** [comment]
Thanks for your interest in the project. Previously we had no demand on implementing UDTs. But now we will try to make it. Could you, please provide me with example how you normally use it?

**Comment by hampuskraft** [comment]
Thanks for the quick reply! Here's an example:

```cql
CREATE TYPE placeholder (
  asset text,
  version int
);

CREATE TYPE thumbnail (
  asset text,
  animated boolean,
  placeholder frozen<placeholder>
);

CREATE TABLE users (
  user_id bigint,
  avatar_data frozen<thumbnail>,
  PRIMARY KEY ((user_id))
);
```

I would then treat UDTs as regular Python dicts throughout the code. The official Python driver for Cassandra/Scylla requires using namedtuples instead of dicts (or [registering classes](https://python-driver.docs.scylladb.com/stable/user-defined-types.html), which is required if not using prepared statements). This requires me to do the following before passing them to my Pydantic models, which isn't very ergonomic:

```py
def to_dict(udt: Any) -> dict[str, Any]:
    return {field: getattr(udt, field) for field in udt._fields}

def to_udt(data: dict[str, Any]) -> SimpleNamespace:
    return SimpleNamespace(**data)
```

**Comment by s3rius** [comment]
Maybe adding support for pydantic models and dataclasses would be a good point.

**Comment by hampuskraft** [comment]
Yeah. This library supports automatically converting results using `as_class`, which should work out of the box with nested models (UDTs) when using Pydantic. To work with dataclasses, using `as_class`, one would likely need to add custom logic to the `__post_init__` constructor. Regardless, things should be relatively clean and straightforward if UDTs work similarly to maps in the code since the Scylla Rust Driver provides `field_types` (`Vec<(String, ColumnType)>`) in `UserDefinedType`, allowing one to construct a HashMap/PyDict for serialization/deserialization recursively.

**Comment by s3rius** [comment]
@hampuskraft, UDT support will be released in 1.3.0. Take a look at README.md how to work with UDTs.

**Comment by hampuskraft** [comment]
It works great, thanks!
